packages/core/src/submodules/cbor/cbor-encode.ts

```diff
@@ -0,0 +1,250 @@
+import { fromUtf8 } from "@smithy/util-utf8";
+
+import {
+  CborMajorType,
+  extendedFloat16,
+  extendedFloat32,
+  extendedFloat64,
+  majorList,
+  majorMap,
+  majorNegativeInt64,
+  majorSpecial,
+  majorTag,
+  majorUint64,
+  majorUnstructuredByteString,
+  majorUtf8String,
+  specialFalse,
+  specialNull,
+  specialTrue,
+  Uint64,
+} from "./cbor-types";
+import { alloc } from "./cbor-types";
+
+const USE_BUFFER = typeof Buffer !== "undefined";
+
+type BufferWithUtf8Write = Buffer & {
+  utf8Write(str: string, index: number): number;
+};
+
+const initialSize = 2048;
+let data: Uint8Array = alloc(initialSize);
+let dataView: DataView = new DataView(data.buffer, data.byteOffset, data.byteLength);
+let cursor: number = 0;
+
+function ensureSpace(bytes: number) {
+  const remaining = data.byteLength - cursor;
+  if (remaining < bytes) {
+    if (cursor < 16_000_000) {
+      resize(Math.max(data.byteLength * 4, data.byteLength + bytes));
+    } else {
+      resize(data.byteLength + bytes + 16_000_000);
+    }
+  }
+}
+
+export function toUint8Array(): Uint8Array {
+  const out = alloc(cursor);
+  out.set(data.subarray(0, cursor), 0);
+  cursor = 0;
+  return out;
+}
+
+export function resize(size: number) {
+  const old = data;
+  data = alloc(size);
```

Comment:
The global encoding buffer is inaccessible to consumers, and so it cannot be passed to workers and won't be subject to concurrent operations. Workers will each have a copy of that buffer. 

Because the encoding function is synchronous and the runtime single-threaded, the buffer is in use by only one encoding process at a time. A copy is made when the result is handed over to the outgoing request. This copy is done very quickly and represents <1% of the time spent.

I believe using a pool of buffers would not improve the speed or memory usage, even if it could avoid that copy operation, due to the speed of that copy operation and the overhead of managing the buffer pool.